src/docfx/config/ops/OpsCredentialHandler.cs

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Net.Http;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Identity;
+using Azure.Security.KeyVault.Secrets;
+
+namespace Microsoft.Docs.Build
+{
+    internal class OpsCredentialHandler : DelegatingHandler
+    {
+        private const string DocsOPSTokenHeader = "X-OP-BuildUserToken";
+
+        private static readonly SecretClient s_secretClient = new(new("https://docfx.vault.azure.net"), new DefaultAzureCredential());
+
+        private static readonly Lazy<Task<string>> s_opsTokenProd = new(() => GetSecret("OpsBuildTokenProd"));
+        private static readonly Lazy<Task<string>> s_opsTokenSandbox = new(() => GetSecret("OpsBuildTokenSandbox"));
+
+        public OpsCredentialHandler(HttpMessageHandler innerHandler)
+            : base(innerHandler)
+        {
+        }
+
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            // Default header which allows fallback to public data when credential is not provided.
+            request.Headers.TryAddWithoutValidation("X-OP-FallbackToPublicData", true.ToString());
+            await FillOpsToken(request);
+            return await base.SendAsync(request, cancellationToken);
+        }
+
+        private static async Task FillOpsToken(HttpRequestMessage request, DocsEnvironment? environment = null)
+        {
+            if (IsSameDocsEnvironmentRequest(request))
```

Comment:
This line makes me want to get away from using `DelegatingHandler` for middlewares. Its contract makes passing variables like `environment` here so much harder.

All we need is to turn the existing credential provider from `Action<HttpRequestMessage>` to:

```csharp
delegate Task<HttpResponseMessage> CredentialProvider(HttpRequestMessage request, Func<HttpRequestMessage, Task<HttpResponseMessage>> next);
```

Then you only need to change places that calls `_credentialProvider?.Invoke(request);`

Reply:
Then the _credentialProvider.invoke need to be called everywhere, and it looks wired to keep a state in the delegate when we need to keep the refreshedToken.